src/main/java/de/tum/in/www1/artemis/domain/hestia/CoverageReport.java

```diff
@@ -0,0 +1,64 @@
+package de.tum.in.www1.artemis.domain.hestia;
+
+import java.util.Set;
+
+import javax.persistence.*;
+
+import org.hibernate.annotations.Cache;
+import org.hibernate.annotations.CacheConcurrencyStrategy;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+import de.tum.in.www1.artemis.domain.DomainObject;
+import de.tum.in.www1.artemis.domain.ProgrammingSubmission;
+
+/**
+ * A testwise coverage report representing the executed code by file path of a single ProgrammingExerciseTestCase.
+ * The entries contain the information about executed code by the start line and the length (i.e. number of lines) of
+ * a consecutively executed block.
+ */
+@Entity
+@Table(name = "coverage_report")
+@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class CoverageReport extends DomainObject {
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "submission_id", referencedColumnName = "id")
+    @JsonIgnore
+    private ProgrammingSubmission submission;
+
+    // when retrieving only the aggregated data (such as covered line ratio), we do not need the file reports
+    @OneToMany(mappedBy = "fullReport", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
+    @JsonIgnoreProperties("fullReport")
+    private Set<CoverageFileReport> fileReports;
+
+    @Column(name = "covered_line_ratio")
+    private Double coveredLineRatio;
```

Comment:
Is this the coverage of the whole submission? Basically we want to measure the coverage for diff between template and solution, right? 

If yes, I would document this and also make sure to define how you store the value. Is 99.5% stored as 99.5 or as 0.995 e.g.?

And does that mean, that the coverage is always stored for programming submissions of the solution participation? Or can it also be stored for template participations? (I think this needs to be documented well in the Java class ProgrammingSubmission)

Reply:
This value refers to the "isolated" coverage for the solution participations submission. The value is in the range of [0, 1].

The coverage can not be recorded for template submissions, because it can only be recorded if all tests succeed - the opposite of what is wanted for the template. Calculating the Coverage-Diff will be part of a Follow-Up-PR.

I added JavaDoc concerning this.